### DISCUSSION TITLE
Biplanar Gradient Generation Error

### DISCUSSION THREAD

**Discussion by Thaejaesh** [kev-m/pyCoilGen#70] [Q&A] [answered]
Hi All,

Thanks for making such a good tool, it has been great so far. I have been trying to generate a biplanar x-gradient based off of the example code but keep running into the same error despite change multiple parameters. 
The error is:

```
Load geometry:
Using plugin:  create_bi_planar_mesh
Split the mesh and the stream function into disconnected pieces.
Upsample the mesh by subdivision:
Parameterize the mesh:
Define the target field:
Calculate mesh one ring:
Create the basis function container which represents the current density:
Calculate the sensitivity matrix:
Calculate the gradient sensitivity matrix:
Calculate the resistance matrix:
Optimize the stream function toward target field and secondary constraints:
INFO:pyCoilGen.sub_functions.stream_function_optimization:Optimising with Tikhonov regularisation.
Calculate the potential levels for the discretization:
Generate the contours:
Process contours: Evaluate loop significance
Find the minimal distance between the contour lines:
Group the contour loops in topological order:
  -- Part 0 has 46 topological groups
  -- Part 1 has 46 topological groups
Calculate center locations of groups:
Interconnect the single groups:
ERROR:__main__:Caught exception: Opening of loop not possible, no overlap between cut sphere and loop
Traceback (most recent call last):
  File "/Users/tj/Documents/Code/Python/LibTests/pyCoilGen/examples/test_cyl_z.py", line 51, in <module>
    result = pyCoilGen(log, arg_dict)
  File "/opt/homebrew/lib/python3.10/site-packages/pyCoilGen/pyCoilGen_release.py", line 395, in pyCoilGen
    raise e
  File "/opt/homebrew/lib/python3.10/site-packages/pyCoilGen/pyCoilGen_release.py", line 307, in pyCoilGen
    coil_parts = interconnect_within_groups(coil_parts, input_args)  # 15
  File "/opt/homebrew/lib/python3.10/site-packages/pyCoilGen/sub_functions/interconnect_within_groups.py", line 124, in interconnect_within_groups
    opened_loop, uv, _ = open_loop_with_3d_sphere(
  File "/opt/homebrew/lib/python3.10/site-packages/pyCoilGen/sub_functions/open_loop_with_3d_sphere.py", line 56, in open_loop_with_3d_sphere
    raise ValueError("Opening of loop not possible, no overlap between cut sphere and loop")
ValueError: Opening of loop not possible, no overlap between cut sphere and loop
```

I have tried a few renditions of messing with the arguments, here is my current attempt

```

arg_dict = {
        'field_shape_function': 'x',  # definition of the target field
        'coil_mesh': 'create bi-planar mesh',
        'biplanar_mesh_parameter_list': [0.25, 0.25, 20, 20, 1, 0, 0, 0, 0, 0, 0.2],
        'target_region_radius': 0.1,  # in meter
        
        'surface_is_cylinder_flag': False,
        'make_cylindrical_pcb': False,
        
        
        'pot_offset_factor': 0.25,
        'levels': 15,
        # 'circular_diameter_factor': 0.05,
        'target_region_radius': 0.1,  # in meter
        'interconnection_cut_width': 0.025,
        'conductor_cross_section_width': 0.002, 
        'min_loop_significance': 3,
        'normal_shift_length': 0.005,
        'iteration_num_mesh_refinement': 1,  # the number of refinements for the mesh;
        'force_cut_selection': ['high'],
        'level_set_method': 'primary',
        
        'skip_postprocessing': False,
        'skip_inductance_calculation': True,
        'tikhonov_reg_factor': 10,  # Tikhonov regularization factor for the SF optimization
        
        
        'save_stl_flag': True,
        'output_directory': 'images',  # [Current directory]
        'project_name': 'grad_Z',
        'persistence_dir': 'debug',
        'debug': DEBUG_BASIC,
    }

    result = pyCoilGen(log, arg_dict)

```

Thanks in advance!

**Comment by kev-m** [comment]
Hello,
Unfortunately, I am not an MRI coil engineer, so I can only make guesses as to what is going wrong. However, it is clear to me that something is wrong with the configuration - there are way too many topological groups in the output. This tells me that the algorithm in unable to find a meaningful solution. 
I suggest that you draw out the expected design on paper and confirm that your target field is where you expect it, and that the coil mesh planes are oriented as you expect. Basically, double check the layout that you are describing by visualising it somehow.
Otherwise, perhaps @Philipp-MR might have some further suggestions.

**Comment by kev-m** [comment]
Note that in the original bi-planar mesh, the example loads the mesh from an existing 3D STL file, which is rotated 90 degrees about the Z-axis compared to the "create bi-planar mesh" mesh builder.

The STL file:
![bi_planar_rectangles_image](https://github.com/kev-m/pyCoilGen/assets/8843655/cc3ce149-33e2-4445-bfd9-e5bca806e3ab)

Created with "create bi-planar mesh":
![create bi-planar mesh](https://pycoilgen.readthedocs.io/en/stable/_images/builder_biplanar_mesh.png)

Changing your gradient field from "x" to "y" might give you a meaningful result. Or rotate the built mesh to be parallel to the STL mesh.

**Reply by Thaejaesh** [reply]
Hi Kev-m,

Thanks for the response! Yea, I have looked into the difference between the pre-created STL and the mesh builder, I have tried setting the gradient to "y"(just re-ran to make sure) and it has not changed the output. I had noticed the difference to the STL and had tried recreating the example STL using the mesh builder and have yet to be able to get it to work despite all other params being the same.
Maybe there is a parameter I have wrong or am missing, I have been pretty set on bi-planar since it makes prototyping easier for me but maybe I just proceed with cylindrical?

Thanks,
Tj

**Comment by kev-m** [comment]
Hello Tj,

I'm embarrassed to admit that you have actually helped me uncover a bug in the `build_biplanar_mesh.py` factory function. 
The constructor was using the wrong data from the `input_args` structure. 

Instead of:
```python
    mesh_data = build_biplanar_mesh(*input_args.biplanar_mesh_parameter_list)
```
it should read:
```python
    mesh_data = build_biplanar_mesh(*input_args.biplanar_mesh_parameter_list)
```
Note the ***bi**planar* instead of *planar*.

I am writing a regression test and will make a bugfix release soon.

Thanks for bringing this to my attention.

Finally, as I pointed out before, to emulate the STL file example, you need to change the normal from `[1,0,0]` to `[0,1,0]`, e.g. to:
```
        'biplanar_mesh_parameter_list': [0.25, 0.25, 20, 20, 0, 1, 0, 0, 0, 0, 0.2],
```

I hope this helps you to make progress with the project.

Thanks, 
Kevin

**Comment by kev-m** [comment]
Actually, the parameters that you provided, above, I think will give you problems. 

I believe that the coil mesh is too small and too close to the provided target mesh (sphere) volume.

The following works for me, as a copy/translation of the bi-planar STL file example, except using the bi-planar mesh builder:

```python
    arg_dict = {
        'field_shape_function': 'x',  # definition of the target field

        # 'coil_mesh_file': 'bi_planer_rectangles_width_1000mm_distance_500mm.stl',
        'coil_mesh': 'create bi-planar mesh',
        'biplanar_mesh_parameter_list': [1.0, 1.0,   # planar_height, planar_width of the planar mesh.
                                         20, 20,     # num_lateral_divisions, num_longitudinal_divisions
                                         0.0, 1.0, 0.0,    # target_normal_x, target_normal_y, target_normal_z
                                         0, 0, 0,    # center_position_x, center_position_y, center_position_z
                                         0.5],       # plane_distance (`float`): Distance between the two planes.

        'target_mesh_file': 'none',
        'secondary_target_mesh_file': 'none',
        'secondary_target_weight': 0.5,
        'target_region_radius': 0.1,  # in meter
        # 'target_region_resolution': 10,  # MATLAB 10 is the default
        'use_only_target_mesh_verts': False,
        'sf_source_file': 'none',
        # the number of potential steps that determines the later number of windings (Stream function discretization)
        'levels': 14,
        # a potential offset value for the minimal and maximal contour potential ; must be between 0 and 1
        'pot_offset_factor': 0.25,
        'surface_is_cylinder_flag': True,
        # the width for the interconnections are interconnected; in meter
        'interconnection_cut_width': 0.05,
        # the length for which overlapping return paths will be shifted along the surface normals; in meter
        'normal_shift_length': 0.01,
        'iteration_num_mesh_refinement': 0, ### 1,  # the number of refinements for the mesh;
        'set_roi_into_mesh_center': True,
        'force_cut_selection': ['high'],
        # Specify one of the three ways the level sets are calculated: "primary","combined", or "independent"
        'level_set_method': 'primary',
        'skip_postprocessing': False,
        'skip_inductance_calculation': False,
        'tikhonov_reg_factor': 10,  # Tikhonov regularization factor for the SF optimization

        'output_directory': 'images/issue70',  # [Current directory]
        'project_name': 'biplanar_xgradient_Tj',
        'persistence_dir': 'debug',
        'debug': DEBUG_BASIC,
    }
```

Note that I reduced `iteration_num_mesh_refinement` from 1 to 0, in order to speed up computation, to allow me to quickly verify that the coil path makes sense.

**Reply by Thaejaesh** [reply]
Hi Kevin,

Thanks for the updates, for some reason I still get the same error even when copy-pasting your arg_dicts above. I will give the library a clean install and retry when I get home from work and see if that changes anything.

Tj

**Reply by kev-m** [reply, marked as answer]
I have just published an updated package now. 

The bugfix is present in version 0.2.2.

**Reply by Thaejaesh** [reply]
Thanks! It is working now!
